source/common/http/http1/parser_impl.h

```diff
@@ -0,0 +1,39 @@
+#pragma once
+
+#include <memory>
+
+#include "common/http/http1/parser.h"
+
+namespace Envoy {
+namespace Http {
+namespace Http1 {
+
+class HttpParserImpl : public Parser {
+public:
+  HttpParserImpl(MessageType type, ParserCallbacks* data);
+  ~HttpParserImpl();
+
+  // Http1::Parser
+  RcVal execute(const char* data, int len) override;
+  void resume() override;
+  ParserStatus pause() override;
+  ParserStatus getStatus() override;
+  uint16_t statusCode() const override;
+  int httpMajor() const override;
+  int httpMinor() const override;
+  absl::optional<uint64_t> contentLength() const override;
+  void hasContentLength(bool val) override;
+  bool isChunked() const override;
+  absl::string_view methodName() const override;
+  absl::string_view errnoName(int rc) const override;
+  int hasTransferEncoding() const override;
+  int statusToInt(const ParserStatus code) const override;
+
+private:
+  class Impl;
+  std::unique_ptr<Impl> impl_;
```

Comment:
Why the impl_ and extra level of unique_ptrs in the parser implementation?

Reply:
Leftover from @derekargueta -- do you know why? Maybe something went wrong in construction/destruction?

Reply:
When trying to get http_parser and llhttp to play nicely together for the transition period where we'll have to support both libraries at runtime, I was running into the issue of colliding symbols between the two libraries, e.g. `#define`'d `HTTP_ERRNO_MAP`, `HTTP_METHOD_MAP`, etc.

So I resorted to using the PImpl pattern to contain the libraries within separate compilation units (transitive inclusion would still hit this issue, e.g. in the factory). Another solution could be patching one of the libraries to change the defined names, but that seemed pretty intrusive. Once http-parser is removed, we wouldn't have this issue.

This was one reason I initially started this with toggling the library via a compile flag, but on code review it was expressed that we want the parser library to be runtime configurable.

Reply:
Would it be possible to add this explanation to the header file and possibly a TODO to remove the Impl once it is possible to do so?

Reply:
Done.